hazelcast/src/main/java/com/hazelcast/spi/BindAddressProvider.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.spi;
+
+import com.hazelcast.config.NetworkConfig;
+
+import java.net.InetSocketAddress;
+
+/**
+ * Allow to customize:
+ * 1. What address Hazelcast will bind to
+ * 2. What address Hazelcast will advertise to other members on which they can bind to
+ * <p>
+ * This is practical in some cloud environments where the default strategy is not yielding good results.
+ *
+ * @see com.hazelcast.instance.DefaultAddressPicker
+ * @see com.hazelcast.config.NetworkConfig#setBindAddressProviderConfig(com.hazelcast.config.BindAddressProviderConfig)
+ */
+public interface BindAddressProvider {
```

Comment:
I personally prefer `MemberAddressProvider` name. Since there are two provided addresses; bind and public.